File: adk-go/internal/airline/repository/flight.go
package repository

import (
	"context"
	"errors"
	"fmt"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	airline "github.com/tokiou/agents-at-scale/internal/airline"
	"github.com/tokiou/agents-at-scale/internal/platform/postgres/sqlc"
)

type FlightRepository struct{ queries *sqlc.Queries }

func NewFlightRepository(queries *sqlc.Queries) *FlightRepository {
	return &FlightRepository{queries: queries}
}

func (r *FlightRepository) GetByID(ctx context.Context, id uuid.UUID) (*airline.Flight, error) {
	value, err := r.queries.GetFlightByID(ctx, id)
	if errors.Is(err, pgx.ErrNoRows) {
		return nil, ErrFlightNotFound
	}
	if err != nil {
		return nil, fmt.Errorf("get flight: %w", err)
	}
	flight := mapFlight(value)
	return &flight, nil
}

func (r *FlightRepository) SearchAvailable(ctx context.Context, params airline.SearchAvailableFlightsParams) ([]airline.AvailableFlight, error) {
	rows, err := r.queries.SearchAvailableFlights(ctx, sqlc.SearchAvailableFlightsParams{
		Origin: params.Origin, Destination: params.Destination,
		DepartureFrom: pgxTimestamp(params.DepartureFrom), DepartureTo: pgxTimestamp(params.DepartureTo),
		PassengerCount: int32(params.PassengerCount),
	})
	if err != nil {
		return nil, fmt.Errorf("search available flights: %w", err)
	}
	results := make([]airline.AvailableFlight, 0)
	indexes := make(map[uuid.UUID]int)
	for _, row := range rows {
		index, ok := indexes[row.FlightID]
		if !ok {
			index = len(results)
			indexes[row.FlightID] = index
			results = append(results, airline.AvailableFlight{Flight: airline.Flight{
				ID: row.FlightID, FlightNumber: row.FlightNumber, OriginAirport: row.OriginAirport,
				DestinationAirport: row.DestinationAirport, DepartureAt: mapTimestamp(row.DepartureAt),
				ArrivalAt: mapTimestamp(row.ArrivalAt), Status: airline.FlightStatus(row.FlightStatus),
				Capacity: int(row.Capacity), CreatedAt: mapTimestamp(row.FlightCreatedAt),
				UpdatedAt: mapTimestamp(row.FlightUpdatedAt),
			}})
		}
		results[index].Fares = append(results[index].Fares, airline.AvailableFare{
			FareClass: airline.FareClass{ID: row.FareClassID, Code: row.Code, Name: row.Name,
				ChangeAllowed: row.ChangeAllowed, CancellationAllowed: row.CancellationAllowed,
				ChangeFee: row.ChangeFee, CancellationFee: row.CancellationFee,
				CreatedAt: mapTimestamp(row.FareClassCreatedAt)},
			FlightFare: airline.FlightFare{ID: row.FlightFareID, FlightID: row.FlightFareFlightID,
				FareClassID: row.FlightFareClassID, Price: row.Price, Currency: row.FareCurrency,
				AvailableSeats: int(row.AvailableSeats)},
		})
	}
	return results, nil
}

File: langgraph-python/app/airline/repository/flight.py
from uuid import UUID

from sqlalchemy import and_, select
from sqlalchemy.ext.asyncio import AsyncSession, async_sessionmaker

from app.airline.models import FareClass, Flight, FlightFare, FlightStatus
from app.airline.repository.errors import FlightNotFoundError
from app.airline.schemas import (
    AvailableFareSchema,
    AvailableFlightSchema,
    SearchAvailableFlightsParamsSchema,
)


class FlightRepository:
    def __init__(self, sessions: async_sessionmaker[AsyncSession]) -> None:
        self._sessions = sessions

    async def get_by_id(self, flight_id: UUID) -> Flight:
        async with self._sessions() as session:
            flight = await session.scalar(select(Flight).where(Flight.id == flight_id))
        if flight is None:
            raise FlightNotFoundError("flight not found")
        return flight

    async def search_available(
        self, params: SearchAvailableFlightsParamsSchema
    ) -> list[AvailableFlightSchema]:
        statement = (
            select(FlightFare, Flight, FareClass)
            .join(Flight, FlightFare.flight_id == Flight.id)
            .join(FareClass, FlightFare.fare_class_id == FareClass.id)
            .where(
                and_(
                    Flight.origin_airport == params.origin,
                    Flight.destination_airport == params.destination,
                    Flight.departure_at >= params.departure_from,
                    Flight.departure_at <= params.departure_to,
                    Flight.status == FlightStatus.SCHEDULED,
                    FlightFare.available_seats >= params.passenger_count,
                )
            )
            .order_by(Flight.departure_at, FlightFare.price)
        )
        async with self._sessions() as session:
            rows = (await session.execute(statement)).all()
        results: dict[UUID, AvailableFlightSchema] = {}
        for flight_fare, flight, fare_class in rows:
            result = results.setdefault(
                flight.id,
                AvailableFlightSchema(flight=flight, fares=[]),
            )
            result.fares.append(
                AvailableFareSchema(fare_class=fare_class, flight_fare=flight_fare)
            )
        return list(results.values())
